src/amber/pipes/cors.cr

```
@@ -1,81 +1,133 @@
-        max_age
+      alias OriginType = Array(String | Regex)
+      FORBIDDEN = "Forbidden for invalid origins, methods or headers"
+      ALLOW_METHODS = %w(PUT PATCH DELETE)
+      ALLOW_HEADERS = %w(Accept Content-type)
 
-      ALLOW_METHODS = %w(GET HEAD POST DELETE OPTIONS PUT PATCH)
-      ALLOW_HEADERS = %w(accept content-type)
+      property origins, headers, methods, credentials, max_age
+      @origin : Origin
 
       def initialize(
-        @allow_origin = "*",
-        @allow_methods = ALLOW_METHODS,
-        @allow_headers = ALLOW_HEADERS,
-        @allow_credentials = false,
-        @max_age = 0
+        @origins : OriginType = ["*", %r()],
+        @methods = ALLOW_METHODS,
+        @headers = ALLOW_HEADERS,
+        @credentials = false,
+        @max_age : Int32? = 0,
+        @expose_headers : Array(String)? = nil,
+        @vary : String? = nil
       )
+        @origin = Origin.new(origins)
       end
 
-      def initialize(
-        @allow_origin = "*",
-        allow_methods : String = ALLOW_METHODS.join(", "),
-        allow_headers : String = ALLOW_HEADERS.join(", "),
-        @allow_credentials = false,
-        @max_age = 0
-      )
-        @allow_methods = allow_methods.strip.split /[\s,]+/
-        @allow_headers = allow_headers.strip.split /[\s,]+/
+      def call(context : HTTP::Server::Context)
+        if @origin.match?(context.request)
+          put_expose_header(context.response)
+          Preflight.request?(context, self)
+          put_response_headers(context.response)
+          call_next(context)
+        else
+          return forbidden(context)
+        end
       end
 
-      def call(context : HTTP::Server::Context)
-        context.response.headers["Access-Control-Allow-Origin"] = allow_origin
+      def forbidden(context)
+        context.response.headers["Content-Type"] = "text/plain"
+        context.response.respond_with_error FORBIDDEN, 403
```

How will this interact with the Error pipe?

---

This writes to the response without raising an error. Should we raise an error here?

---

I think so, that way it can be caught by the errors pipe and rendered nicely

---

So we do not need to raise error here just respond with 403 according to the CORS specs since the Pref-light request is done behind the scenes by the browser